### PR TITLE
Sync one ingress

### PR DESCRIPTION
**Pull request by bowei**
A bunch of patches:
* A lot of clean up and movement
* Sync one ingress (as opposed to all) in Checkpoint
* Call GC explicitly when an Ingress is deleted
* Shutdown() firewall when there are no more Ingresses

### CODE REVIEW

**Comment by bowei** [comment]
@G-Harmon

**Comment by bowei** [comment]
Look at this change on a commit wise basis, otherwise will be hard to review.

File: pkg/controller/cluster_manager.go [hunk 1]
```diff
@@ -172,15 +174,17 @@ func (c *ClusterManager) GC(lbNames []string, nodePorts []backends.ServicePort)
...
 		igName := c.ClusterNamer.InstanceGroup()
 		glog.Infof("Deleting instance group %v", igName)
-		igErr = c.instancePool.DeleteInstanceGroup(igName)
+		if err := c.instancePool.DeleteInstanceGroup(igName); err != err {
+			return err
+		}
 	}
-	if igErr != nil {
-		return igErr
+	if len(lbNames) == 0 {
```
Comment (freehan) [inline pkg/controller/cluster_manager.go]: merge it with the previous if statement?

Add a comment to say if there is no lbs. Remove instance group and firewall

File: pkg/controller/controller.go [hunk 2]
```diff
@@ -92,26 +92,24 @@ type LoadBalancerController struct {
...
+	broadcaster := record.NewBroadcaster()
+	broadcaster.StartLogging(glog.Infof)
+	broadcaster.StartRecordingToSink(&unversionedcore.EventSinkImpl{
 		Interface: kubeClient.Core().Events(""),
 	})
 	lbc := LoadBalancerController{
 		client:              kubeClient,
+		ctx:                 ctx,
+		ingLister:           StoreToIngressLister{Store: ctx.IngressInformer.GetStore()},
+		nodeLister:          ctx.NodeInformer.GetIndexer(),
```
Comment (freehan) [inline pkg/controller/controller.go]: Probably makes sense to throw nodeLister and getReadyNodes into Translator.

File: pkg/context/context.go [hunk 3]
```diff
@@ -19,35 +19,88 @@ package context
...
+	kubeClient kubernetes.Interface
+
 	IngressInformer  cache.SharedIndexInformer
 	ServiceInformer  cache.SharedIndexInformer
 	PodInformer      cache.SharedIndexInformer
 	NodeInformer     cache.SharedIndexInformer
 	EndpointInformer cache.SharedIndexInformer
+
+	// Map of namespace => record.EventRecorder.
+	recorders map[string]record.EventRecorder
```
Comment (freehan) [inline pkg/context/context.go]: EventRecorder per namespace?

File: pkg/controller/node.go [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+	"time"
+
+	listers "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/ingress-gce/pkg/context"
+	"k8s.io/ingress-gce/pkg/utils"
+)
+
+const (
+	nodeResyncPeriod = 1 * time.Second
```
Comment (nicksardo) [inline pkg/controller/node.go]: Resync every second?

File: pkg/controller/utils.go [hunk 5]
```diff
@@ -206,3 +208,19 @@ func uniq(nodePorts []backends.ServicePort) []backends.ServicePort {
 	}
 	return nodePorts
 }
+
+// getReadyNodeNames returns names of schedulable, ready nodes from the node lister.
+func getReadyNodeNames(lister listers.NodeLister) ([]string, error) {
+	nodeNames := []string{}
```
Comment (nicksardo) [inline pkg/controller/utils.go]: I know you didn't write this, but
Nit: var nodeNames []string

File: pkg/context/context.go [hunk 6]
```diff
@@ -51,6 +51,26 @@ func NewControllerContext(kubeClient kubernetes.Interface, namespace string, res
 	return context
 }
 
+// HasSynced returns true if all relevant informers has been synced.
+func (ctx *ControllerContext) HasSynced() bool {
+
```
Comment (nicksardo) [inline pkg/context/context.go]: Extra space

File: pkg/controller/cluster_manager.go [hunk 7]
```diff
@@ -105,6 +105,8 @@ func (c *ClusterManager) shutdown() error {
 // If in performing the checkpoint the cluster manager runs out of quota, a
 // googleapi 403 is returned.
 func (c *ClusterManager) Checkpoint(lbs []*loadbalancers.L7RuntimeInfo, nodeNames []string, backendServicePorts []backends.ServicePort, namedPorts []backends.ServicePort, firewallPorts []int64) ([]*compute.InstanceGroup, error) {
+	glog.V(4).Infof("Checkpoint %q, len(lbs)=%v, len(nodeNames)=%v, lne(backendServicePorts)=%v, len(namedPorts)=%v, len(firewallPorts)=%v", len(lbs), len(nodeNames), len(backendServicePorts), len(namedPorts), len(firewallPorts))
```
Comment (nicksardo) [inline pkg/controller/cluster_manager.go]: Provided only 5 values - expected 6.

Typo: "lne(backendServicePorts)"

File: pkg/context/context.go (hunk 3 above)
Comment (nicksardo) [reply pkg/context/context.go]: I'm a bit confused by this too. The service controller doesn't specify the namespace:

https://github.com/kubernetes/kubernetes/blob/e35d67e032e6a064c0f867325ea2a48bca1ef9cf/pkg/controller/service/service_controller.go#L118

File: pkg/controller/translator/translator_test.go [hunk 8]
```diff
@@ -0,0 +1,317 @@
+/*
+Copyright 2017 The Kubernetes Authors.
```
Comment (G-Harmon) [inline pkg/controller/translator/translator_test.go]: s/2017/2018/

File: pkg/controller/node.go [hunk 9]
```diff
@@ -0,0 +1,72 @@
+/*
+Copyright 2017 The Kubernetes Authors.
```
Comment (G-Harmon) [inline pkg/controller/node.go]: 2018

File: pkg/controller/node.go [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+)
+
+const (
+	nodeResyncPeriod = 1 * time.Second
+)
+
+// NodeController synchronizes the state of the nodes to the unmanaged instance
+// groups.
+type NodeController struct {
+	lister cache.Indexer
```
Comment (G-Harmon) [inline pkg/controller/node.go]: newbie comment: is it worth commenting members? Is there style guidance on always commenting struct/class member variables? (In google c++, we say we should.)

File: pkg/controller/translator/translator.go [hunk 11]
```diff
@@ -342,9 +342,10 @@ func (t *GCE) GatherFirewallPorts(svcPorts []backends.ServicePort, includeDefaul
 	if includeDefaultBackend {
 		svcPorts = append(svcPorts, *t.bi.DefaultBackendNodePort())
 	}
+
 	portMap := map[int64]bool{}
 	for _, p := range svcPorts {
-		if p.NEGEnabled {
+		if t.negEnabled && p.NEGEnabled {
```
Comment (G-Harmon) [inline pkg/controller/translator/translator.go]: Is this a bug fix? If so, should it go in its own commit?

File: pkg/controller/cluster_manager.go (hunk 1 above)
Comment (bowei) [reply pkg/controller/cluster_manager.go]: done

File: pkg/firewalls/firewalls.go [hunk 12]
```diff
@@ -54,6 +54,7 @@ func NewFirewallPool(cloud Firewall, namer *utils.Namer) SingleFirewallPool {
 
 // Sync sync firewall rules with the cloud.
 func (fr *FirewallRules) Sync(nodePorts []int64, nodeNames []string) error {
+	glog.V(4).Infof("Sync(%v, %v)", nodePorts, nodeNames)
```
[the remark this answers is not in the document]
Comment (bowei) [reply pkg/firewalls/firewalls.go]: It should be clear from the filename in the log line

File: pkg/controller/node.go (hunk 4 above)
Comment (bowei) [reply pkg/controller/node.go]: ? what do you want me to do here? This is the existing policy.

File: pkg/controller/utils.go (hunk 5 above)
Comment (bowei) [reply pkg/controller/utils.go]: done

File: pkg/context/context.go (hunk 3 above)
Comment (bowei) [reply pkg/context/context.go]: Let's keep this for now, I'll remove in a later commit once we resolve the unit test log spam

File: pkg/context/context.go (hunk 6 above)
Comment (bowei) [reply pkg/context/context.go]: done

File: pkg/controller/cluster_manager.go (hunk 7 above)
Comment (bowei) [reply pkg/controller/cluster_manager.go]: done

File: pkg/controller/node.go (hunk 9 above)
Comment (bowei) [reply pkg/controller/node.go]: We typically don't update the year as it causes too much spam.

File: pkg/controller/node.go (hunk 10 above)
Comment (bowei) [reply pkg/controller/node.go]: good point, yes in general unless extremely obvious.

File: pkg/controller/node.go [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+func NewNodeController(ctx *context.ControllerContext, cm *ClusterManager) *NodeController {
+	c := &NodeController{
+		lister: ctx.NodeInformer.GetIndexer(),
+		cm:     cm,
+	}
+	c.queue = utils.NewPeriodicTaskQueue("nodes", c.sync)
+
+	ctx.NodeInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    c.queue.Enqueue,
+		DeleteFunc: c.queue.Enqueue,
```
Comment (nicksardo) [inline pkg/controller/node.go]: Can we add a TODO here that we need to improve this NodeController? 

We don't react on the UpdateFunc because it's supposedly very noisy. However, when new nodes are added, they are typically NotReady. This means in practice that instance groups are synced when the NodeInformer re-calls AddFunc (every 10 minutes).

File: pkg/controller/node.go (hunk 4 above)
Comment (nicksardo) [reply pkg/controller/node.go]: Could we rename to something else? Looks like this is just the frequency that the PeriodicTaskQueue checks the queue, not at which items in the queue are re-enqueued. Am I mistaken?

File: pkg/context/context.go (hunk 3 above)
Comment (nicksardo) [reply pkg/context/context.go]: SGTM

File: pkg/controller/node.go (hunk 13 above)
Comment (bowei) [reply pkg/controller/node.go]: done

File: pkg/controller/node.go (hunk 4 above)
Comment (bowei) [reply pkg/controller/node.go]: I changed to nodeUpdatePeriod

File: pkg/controller/node.go (hunk 9 above)
Comment (G-Harmon) [reply pkg/controller/node.go]: This isn't an update. This is a new file. (I agree for changes to files the copyright isn't normally changed.)
